### ISSUE TITLE
Missing `Navigator.clipboard` (clipboard asynchronous API)

### ISSUE THREAD

**Issue by modrzew** [Microsoft/TypeScript#26728]
**TypeScript Version:**  3.0.0-dev.20180711

**Search Terms:** navigator, clipboard

**Code**

```ts
console.log(window.navigator.clipboard)
```

**Expected behavior:** No error.

**Actual behavior:**
```
error TS2339: Property 'clipboard' does not exist on type 'Navigator'
```

**Playground Link:** http://www.typescriptlang.org/play/#src=window.navigator.clipboard%3B

**Related Issues:** none found

W3C: https://www.w3.org/TR/clipboard-apis/#async-clipboard-api
MDN: https://developer.mozilla.org/en-US/docs/Web/API/Clipboard

**Comment by MartinJohns** [comment]
Probably missing because it's still a working draft and no browser supports it yet: https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#Browser_compatibility

Why add something to the standard library that is not supported?

**Comment by modrzew** [comment]
> no browser supports it yet

It's available in Chrome since 62 (behind a flag) and enabled by default since 66: https://www.chromestatus.com/feature/5861289330999296

**Comment by RyanCavanaugh** [comment]
We don't add non-standard DOM APIs until they're at least widely adopted.

**Comment by glebm** [comment]
It is an [experimental API](https://www.w3.org/TR/clipboard-apis/). However, `readText` and `writeText` are also [supported in Firefox 63+](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard).

As a workaround:

```typescript
// lib.dom.clipboard.d.ts

// Type declarations for Clipboard API
// https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
interface Clipboard {
  writeText(newClipText: string): Promise<void>;
  // Add any other methods you need here.
}

interface NavigatorClipboard {
  // Only available in a secure context.
  readonly clipboard?: Clipboard;
}

interface Navigator extends NavigatorClipboard {}
```

**Comment by lucifer1004** [comment]
> It is an [experimental API](https://www.w3.org/TR/clipboard-apis/). By the way, `readText` and `writeText` are also [supported in Firefox 63+](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#Browser_compatibility).
> 
> As a workaround:
> 
> ```ts
> // navigator.clipboard.d.ts
> 
> // Type declarations for Clipboard API
> // https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
> interface Clipboard {
>   writeText(newClipText: string): Promise<void>;
>   // Add any other methods you need here.
> }
> 
> interface NavigatorClipboard {
>   // Only available in a secure context.
>   readonly clipboard?: Clipboard;
> }
> 
> interface Navigator extends NavigatorClipboard {}
> ```

Where should I place this file?

**Comment by maciaszczykm** [comment]
Hey, isn't it a good moment to go back to this issue and reopen it? I would really love to have it to avoid any workarounds. Chrome supports it from v66 now I have v71. Should be pretty stable.

**Comment by blixt** [comment]
How is "widely adopted" defined? It seems to me that according to http://gs.statcounter.com/ over half of all desktop users were on a Chrome version that supported this API already for months when the comment "We don't add non-standard DOM APIs until they're at least widely adopted." was made.

**Comment by maciaszczykm** [comment]
@RyanCavanaugh Can you give us any answer?

**Comment by cpplearner** [comment]
https://github.com/Microsoft/TypeScript/blob/2f9218f/src/lib/dom.generated.d.ts#L10489

**Comment by HighSoftWare96** [comment]
> > It is an [experimental API](https://www.w3.org/TR/clipboard-apis/). By the way, `readText` and `writeText` are also [supported in Firefox 63+](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard#Browser_compatibility).
> > As a workaround:
> > ```ts
> > // navigator.clipboard.d.ts
> > 
> > // Type declarations for Clipboard API
> > // https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
> > interface Clipboard {
> >   writeText(newClipText: string): Promise<void>;
> >   // Add any other methods you need here.
> > }
> > 
> > interface NavigatorClipboard {
> >   // Only available in a secure context.
> >   readonly clipboard?: Clipboard;
> > }
> > 
> > interface Navigator extends NavigatorClipboard {}
> > ```
> 
> Where should I place this file?

I've done that inline like this:
```
// !! in any place or file !!
// Type declarations for Clipboard API
// https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
interface Clipboard {
  writeText(newClipText: string): Promise<void>;
  // Add any other methods you need here.
}
interface NavigatorClipboard extends Navigator {
  // Only available in a secure context.
  readonly clipboard?: Clipboard;
}
interface NavigatorExtended extends NavigatorClipboard {}
```

And where you need to use the clipboard you can cast the navigator to the created interface:
```
    (navigator as NavigatorExtended).clipboard
      .writeText(event.target.innerText)
      .then(() => {
        alert("Copied!");
      }); 
```
(if the extended interface is in an external file you need to import it first!)

**Comment by PaulieScanlon** [comment]
Perhaps not a great workaround but i've done this:

// type new var with any
`let anyNavigator: any`
// assign new var to window.navigator
`anyNavigator = window.navigator`

// use new var in place of navigator
`anyNavigator.clipboard.writeText(event.target.defaultValue)`

seems to work ok with `"typescript": "^3.2.4"`

**Comment by rassie** [comment]
So basically, this is fixed since [this commit](https://github.com/microsoft/TypeScript/commit/2f9218f346bbc606cccc4aebfca1c2c4f3167fee) and should be available since 3.4?

**Comment by fr0** [comment]
Both Chrome and Firefox support the clipboard `read` and `write` methods now . Those should be added to the type definitions (which currently only contain `readText` and `writeText`).

**Comment by MartinJohns** [comment]
@fr0 But it's still only a `Working Draft`.

**Comment by 1valdis** [comment]
For me it's a shame that I can't properly play with modern (or experimental) APIs using only standard Typescript definitions, and I have to declare them myself instead.

**Comment by vancityhuddy** [comment]
@1valdis it's very reasonable that Typescript would not include these tools by default. Typescript is targeted to a very wide ranging audience. Giving bleeding-edge tools without warning and by default to developers who are unaware they might be bleeding edge easily leads to unknown bugs. TS is designed to prevent exactly these kinds of issues.

If you're keen to use experimental tools, then you should also be aware of how to extend your definition files to allow them, and how to support users who do not have the latest browsers.

**Comment by gayanDeegala** [comment]
> @1valdis it's very reasonable that Typescript would not include these tools by default. Typescript is targeted to a very wide ranging audience. Giving bleeding-edge tools without warning and by default to developers who are unaware they might be bleeding edge easily leads to unknown bugs. TS is designed to prevent exactly these kinds of issues.
> 
> If you're keen to use experimental tools, then you should also be aware of how to extend your definition files to allow them, and how to support users who do not have the latest browsers.

Can anyone please share how to do this?

**Comment by DominicTobias** [comment]
> @1valdis it's very reasonable that Typescript would not include these tools by default. Typescript is targeted to a very wide ranging audience. Giving bleeding-edge tools without warning and by default to developers who are unaware they might be bleeding edge easily leads to unknown bugs. TS is designed to prevent exactly these kinds of issues.
> 
> If you're keen to use experimental tools, then you should also be aware of how to extend your definition files to allow them, and how to support users who do not have the latest browsers.

I wouldn't exactly describe it as bleeding edge

![Screenshot 2020-08-31 at 02 28 37](https://user-images.githubusercontent.com/760314/91666730-d05e7e80-eb31-11ea-9789-093b062178b2.png)

**Comment by pbadenski** [comment]
This should be sufficient (it will work using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)). Put it in a `clipboard.d.ts` file at the top level under on of your type root directories (tsconfig.json -> typeRoots)

```
interface ClipboardItem {
}

declare var ClipboardItem: {
  prototype: ClipboardItem;
  new(objects: Record<string, Blob>): ClipboardItem;
};

interface Clipboard {
  read(): Promise<Array<ClipboardItem>>;

  write(items: Array<ClipboardItem>): Promise<void>;
}
```

**Comment by aosi87** [comment]
> This should be sufficient (it will work using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)). Put it in a `clipboard.d.ts` file at the top level under on of your type root directories (tsconfig.json -> typeRoots)
> 
> ```
> interface ClipboardItem {
> }
> 
> declare var ClipboardItem: {
>   prototype: ClipboardItem;
>   new(objects: Record<string, Blob>): ClipboardItem;
> };
> 
> interface Clipboard {
>   read?(): Promise<Array<ClipboardItem>>;
> 
>   write?(items: Array<ClipboardItem>): Promise<void>;
> }
> ```

It should be under tsconfig.json -> types, it is a single file (in my case), so no reason to add it into a the rootTypes folders array option.

**Comment by mad4power** [comment]
This still doesn't work in Safari

**Comment by martinboksa** [comment]
Added basic types to `ClipboardItem`
```
interface ClipboardItem {
  readonly types: string[]
  getType: (type: string) => Promise<Blob>
}

declare var ClipboardItem: {
  prototype: ClipboardItem
  new (objects: Record<string, Blob>): ClipboardItem
}

interface Clipboard {
  read?(): Promise<Array<ClipboardItem>>
  write?(items: Array<ClipboardItem>): Promise<void>
}
```
